### ISSUE TITLE
Handling out of order packets

### ISSUE THREAD

**Issue by pavlik-uhuhu** [alpartis/rtp.jitter#2]
Hello, i'm tried to use this jitter buffer for handling out-of-order and dropped packets.
Case is following: when buffer is empty, it will simply accumulate all RTP packets for specified period of time (_nominal_depth_ms), and for that period of time all requests to pop() method will return **BUFFERING** result and internal state will not be changed;
so while it's accumulating, all newly arrival RTP packets will not be reordered (in case if packets were out-of-order) because of the following (`rtp_jitter.cpp/RTPJitter::push`):
```
if (_buffer.size() == 1) {
    _first_buf_sequence = _last_pop_sequence = rtp_sequence;
 }
```
so when first packet arrive - it'll simply set **_first_buf_sequence** and **_last_pop_sequence** to be the same, and while buffer remains in that state, all subsequent calls to push() method will simply add new packets to the end of the buffer:
```
if ((rtp_sequence >= _last_buf_sequence)
         || ((rtp_sequence == 0) && (_last_buf_sequence == UINT16_MAX))
         || (_last_pop_sequence == _first_buf_sequence))
        {
            _buffer.push_back(p);
``` 

So, am i doing something wrong, or this buffer simply doesn't work properly?

**Comment by alpartis** [comment]
Yes, there's a bug there.  Thanks for pointing it out.  That conditional statement should probably take the current buffer size into account so it can catch out-of-order packets while the buffer is filling.  Once the buffer fills, and packets start getting popped, it should be good.  But it looks like out-of-order packets while initially filling the buffer will not be detected and handled properly.

**Comment by pavlik-uhuhu** [comment]
BTW, what does it means "initially filling the buffer"? there is nothing that can stop me from pop-out all packets at once after **nominal_depth_ms** has passed, and reset buffer to be in buffering state again.
So how it even should be used anyway?

**Comment by alpartis** [comment]
You'll probably need to read up a little bit on typical VoIP and RTP applications, but packets don't usually arrive all at once ... usually only one packet about every 20ms (to match the amount of audio in any given packet as it's being recorded on the other end of the conversation, if it's a voice application, for example).  So, if an application that is receiving the RTP audio begins playback as soon as the first packet arrives, there is more likely to be problems during playback if there are any deviations in packet delivery timing.  A jitter buffer is employed to smooth things out and it does this first by accumulating some amount of packets before delivering them to the application for playback -- this is what is meant by "initially filling the buffer."

**Comment by pavlik-uhuhu** [comment]
Well, my application don't playback media it reads from buffer, it just stores media with respect to correct order of packets, so it tries to pop packets out at maximum throughput.
All i was saying, that in this particular implementation of jitter buffer one can easily _reset_ buffer state to be in _buffering state_ in any moment, by simply pop out all packets, so this state's process is not really "initial".

**Comment by alpartis** [comment]
It sounds like your application is not the ideal setting to implement a jitter buffer because you're not concerned with "jitter" ... you're more concerned simply with order.  This rtp.jitter project is more suitable for VoIP applications with near real-time recording, transmission, and playback of audio.  I also wouldn't get hung up on terminology and definitions of "initial."

That said, there is one attribute of _some_ jitter buffer implementations that is not perfectly manifested here: adaptivity.  Without an integral link to an audio codec library, it's impossible to recover from excessive jitter that causes the buffer to have to expand without resorting to simply dropping "good" audio packets from the playback stream.
